src/flood.rs

```diff
@@ -0,0 +1,78 @@
+use crate::view::layer::LayerCoords;
+use crate::view::{Snapshot, View, ViewResource};
+use rgx::color::{Rgba8};
+use rgx::kit::shape2d::{Fill, Rotation, Shape, Stroke};
+use rgx::kit::ZDepth;
+use rgx::rect::Rect;
+
+pub fn flood_fill(
+    view: &View<ViewResource>,
+    starting_point: LayerCoords<u32>,
+    replacement_color: Rgba8,
+) -> Option<Vec<Shape>> {
+    let (snapshot, pixels) = view.current_snapshot(view.active_layer_id)?;
+    let mut canvas = pixels.clone().into_rgba8();
+    let bounds = snapshot.extent.rect();
+    let target_color = get_color(snapshot, &mut canvas, starting_point)?.clone();
+
+    if target_color == replacement_color {
+        return None;
+    }
+
+    let mut points = Vec::new();
+    let mut queue = vec![starting_point];
+
+    while let Some(point) = queue.pop() {
+        if let Some(c) = get_color(snapshot, &mut canvas, point) {
+            if *c == target_color {
+                *c = replacement_color;
+                points.push((point, replacement_color));
+                queue.extend(neighbors(point, bounds));
+            }
+        }
+    }
+
+    Some(to_shapes(points))
+}
+
+fn get_color<'a>(
+    snapshot: &Snapshot,
+    canvas: &'a mut Vec<Rgba8>,
+    point: LayerCoords<u32>,
+) -> Option<&'a mut Rgba8> {
+    let idx = snapshot.layer_coord_to_index(point)?;
+    canvas.get_mut(idx)
+}
+
+fn to_shapes(points: Vec<(LayerCoords<u32>, Rgba8)>) -> Vec<Shape> {
+    let mut rects = Vec::with_capacity(points.len());
+    for (it, color) in points {
+        let x = it.x as f32;
+        let y = it.y as f32;
+        rects.push(Shape::Rectangle(
+            Rect::new(x, y, x + 1.0, y + 1.0),
+            ZDepth::default(),
+            Rotation::ZERO,
+            Stroke::NONE,
+            Fill::Solid(color.into()),
+        ));
```

Comment:
This will be slow on large canvases, but can be improved later.

Reply:
Agreed, it's not ideal. I played around with a scanline-based algorithm that's a lot faster, but the code was pretty hard to read and this worked okay for my small test canvases. I might spend a bit more time trying to optimize things. If we could get it fast enough, we could do some cool things like show a live preview of the pixels we're going to fill.

Reply:
That's neat! If you want to discuss ideas around this, I just setup a community discord server for `rx`: https://discord.gg/VtjUpj8n -- just drop by anytime.